cpp/src/arrow/buffer-builder.h

```diff
@@ -42,7 +42,11 @@ namespace arrow {
 class ARROW_EXPORT BufferBuilder {
  public:
   explicit BufferBuilder(MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT)
-      : pool_(pool), data_(NULLPTR), capacity_(0), size_(0) {}
+      : pool_(pool),
+        data_(/*ensure never null to make ubsan happy and avoid check penalties below*/
+              reinterpret_cast<uint8_t*>(this)),
```

Comment:
Weird choice :-) Might be cleaner to have a static 0-sized array somewhere?

Reply:
Cleaned this up to use MakeNonNull which I factored out.  0-sized static arrays are more challenging then you would expect since they can't be stack allocated according to the C/C++ spec.